contracts/RealmDiamond/facets/ParcelRolesRegistryFacet.sol

```diff
@@ -0,0 +1,235 @@
+        _;
+    }
+
+    modifier onlyValidRole(bytes32 _roleId) {
+        require(s.validRoles[_roleId], "ParcelRolesRegistryFacet: invalid role ID");
+        _;
+    }
+
+    function isValidRole(bytes32 roleId) external view returns (bool) {
+    return s.validRoles[roleId];
+}
+
+    /** External Functions **/
+
+    function grantRole(Role calldata _role) external override  onlyValidRole(_role.roleId){
+        require(_role.expirationDate > block.timestamp, "ParcelRolesRegistryFacet: expiration date must be in the future");
+ 
+        // deposit NFT if necessary
+        // reverts if sender is not approved or original owner
+        address _originalOwner = _depositNft(_role.tokenAddress, _role.tokenId);
+
+        // role must be expired or revocable
+        RoleData storage _roleData = s.erc7432_roles[_role.tokenAddress][_role.tokenId][_role.roleId];
+        require(_roleData.revocable || _roleData.expirationDate < block.timestamp, "ParcelRolesRegistryFacet: role must be expired or revocable");
+      
+        s.erc7432_roles[_role.tokenAddress][_role.tokenId][_role.roleId] = RoleData(
+            _role.recipient,
+            _role.expirationDate,
+            _role.revocable,
+            _role.data
+        );
+
+        emit RoleGranted(
+            _role.tokenAddress,
+            _role.tokenId,
+            _role.roleId,
+            _originalOwner,
+            _role.recipient,
+            _role.expirationDate,
+            _role.revocable,
+            _role.data
+        );
+    }
+
+    function revokeRole(address _tokenAddress, uint256 _tokenId, bytes32 _roleId) external override {
+        address _recipient = s.erc7432_roles[_tokenAddress][_tokenId][_roleId].recipient;
+        address _caller = _getApprovedCaller(_tokenAddress, _tokenId, _recipient);
+
+        // if caller is recipient, the role can be revoked regardless of its state
+        if (_caller != _recipient) {
+            // if caller is owner, the role can only be revoked if revocable or expired
+            require(
+                s.erc7432_roles[_tokenAddress][_tokenId][_roleId].revocable ||
+                    s.erc7432_roles[_tokenAddress][_tokenId][_roleId].expirationDate < block.timestamp,
+                "ParcelRolesRegistryFacet: role is not revocable nor expired"
+            );
+        }
+
+        delete s.erc7432_roles[_tokenAddress][_tokenId][_roleId];
+        emit RoleRevoked(_tokenAddress, _tokenId, _roleId);
+    }
+
+    function unlockToken(address _tokenAddress, uint256 _tokenId) external override {
+        address originalOwner = s.erc7432OriginalOwners[_tokenAddress][_tokenId];
+
+        require(!_hasNonRevocableRole(_tokenAddress, _tokenId), "ParcelRolesRegistryFacet: NFT is locked");
+
+        require(
+            originalOwner == msg.sender || isRoleApprovedForAll(_tokenAddress, originalOwner, msg.sender),
+            "ParcelRolesRegistryFacet: sender must be owner or approved"
+        );
```

Comment:
Invert the order of the `requires`, if you are reading `originalOwner`, you should check access control immediately afterwards